jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalUpcallHandler.java

```diff
@@ -148,36 +150,26 @@ private long getUpcallThunkAddr(MethodHandle target, ResourceScope sessionOrScop
 			nativeSignatureStrs[argLayoutCount] = LayoutStrPreprocessor.getSimplifiedLayoutString(realReturnLayout, false);
 		}
 
-		long addr = 0;
-		synchronized(privateUpcallClassLock) {
-			/* The generated thunks are shared across upcalls or threads only when
-			 * the target method handles are identical within the same session/scope
-			 * given a thunk plus the associated metadata is released automatically
-			 * in native via freeUpcallStub() in OpenJDK when the corresponding
-			 * session/scope is terminated.
-			 */
-			String targetHashScopeStr = target.hashCode() + "#" + sessionOrScope.toString(); //$NON-NLS-1$
-			Integer upcallThunkAddrKey = Integer.valueOf(targetHashScopeStr.hashCode());
-			Long upcallThunkAddr = cachedUpcallThunkAddr.get(upcallThunkAddrKey);
-			if (upcallThunkAddr != null) {
-				addr = upcallThunkAddr.longValue();
-			} else {
-				/* The thunk must be created for each upcall handler given the UpcallMHMetaData object uniquely bound to the thunk
-				 * is only alive for a MemorySession(JDK19)/ResourceScope(JDK17/18) specified in java, which means the upcall handler
-				 * and its UpcallMHMetaData object will be cleaned up automatically once their session/scope is closed.
-				 */
-				metaData = new UpcallMHMetaData(target, sessionOrScope);
-				addr = allocateUpcallStub(metaData, nativeSignatureStrs);
-				cachedUpcallThunkAddr.put(upcallThunkAddrKey, Long.valueOf(addr));
-			}
-		}
-
-		return addr;
+		/* The thunk must be created for each upcall handler given the UpcallMHMetaData object uniquely bound to the thunk
+		 * is only alive for a MemorySession(JDK19)/ResourceScope(JDK17/18) specified in java, which means the upcall handler
+		 * and its UpcallMHMetaData object will be cleaned up automatically once their session/scope is closed.
+		 */
+		metaData = new UpcallMHMetaData(target, sessionOrScope);
+		MHSession targetSession = new MHSession(target, sessionOrScope);
+		return allocateUpcallStub(metaData, nativeSignatureStrs, targetSession.hashCode());
```

Comment:
`targetSession.hashCode()` isn't guaranteed to be unique for all `target/sessionOrScope` pairs.

Reply:
As we discussed offline, there is no feasible way of absolutely avoiding hash collision in the cache even we replace it with `map<MSSession, thunk address>` in java. In addition to that, I don't think it is affordable to remove the stale entry via `UpcallStubs.freeUpcallStub()` in OpenJDK for two reasons:
1)  according to the existing FFI/OpenJDK framework, only the thunk address is visible to both OpenJDK and OpenJ9,
in which case we have to modify everything related to `freeUpcallStub()` in OpenJDK and OpenJ9 to support removing the stale entry. Meanwhile, the cache might need to be defined in OpenJDK rather than OpenJ9 to ensure it can be accessed on both sides. Even so, there is no 100% guarantee that it works out in such case. In addition to that, we might encounter the out-of-sync code related issue if there is any update in OpenJDK.
2) doing a JNI call from native to java via `Java_jdk_internal_foreign_abi_UpcallStubs_freeUpcallStub0` to remove the stale entry in `map<MSSession, thunk address>` in `/jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalUpcallHandler.java` is also unfeasible as it definitely downgrades the overall performance in JVM even though it only happens in the closing stage of the session/scope. The ultimate goal of FFI is to replace JNI in any case rather than leveraging it in the FFI implementation as specified in the JEPs, in which case we can't move forward in this way.

So the only thing we can do for now is to roll back to the original implementation and simply replace the existing cache with `map<MSSession, thunk address>` (but the stale way still remains in the cache after closing the session/scope) if we do think it is way better than the existing cache `cachedUpcallThunkAddr` in `InternalUpcallHandler.java` to guarantee the uniqueness for all `target/sessionOrScope` pairs.

I would suggest to open another issue for the hash collision if we think it does cause problem given this exploration is far away from the intention of this PR (the OOM issue here has nothing to with the hash collision).

@gacholio, @tajila, any suggestion/input on this?

Reply:
Following the discussion with @keithc-ca  offline, I went through the code in OpenJDK (especially in src/java.base/share/classes/jdk/internal/foreign/abi/UpcallStubs.java) against the underlying code in OpenJ9, realizing that there might be a way of removing the stale entry at Java level from invoking a upcall linker method (defined by ourselves in OpenJ9) from OpenJDK to OpenJ9 with the support of the reverse map. So it might be worth trying out to see whether it works in this way.

Reply:
As discussed with @keithc-ca,  it might be problematic to reusing thunks for the same session/scope under the existing OpenJDK framework. In such case, I will remove everything related to the java cache for the short-term solution.